### PR TITLE
coredns-operator using operator-sdk

### PR DESCRIPTION
**Pull request by stealthybox**
This is an alternate implementation, similar to #7.
It addresses #3.

See `README.md` for instructions :)

### Things I focused on:
- Golang native declarations of objects (this was time-consuming)
- Specific reconcile loop for operating CoreDNS:
   - Deleted objects are immediately re-created
   - Toggling `CoreDNS.Spec.ClusterDNS` re-creates Services /w an appropriate ClusterIP
   - Changing `CoreDNS.Spec.Corefile` updates the ConfigMap
      (not hashed for RollingUpdate at the moment, but CoreDNS can reload -- this could be an option)
- Granular operator RBAC using an annotated Role and ClusterRole
- `system-cluster-critical` PriorityClass used only for `kube-system`
- Per-CoreDNS ClusterRole + Binding named by namespace
- Useful `Makefile` /w help from `kind`

### Did not implement:
- Common fields (I did add them to the CoreDNS CRD though)
- Status/Error fields
- YAML base manifests
- Release channels / Versioning
- Generic object patching (ie: NodeAffinity / ResourceRequests)
- Generic enforcement of all Object fields (is there a good way to do a generic diff?)
- Go modules -- current patch is using generated dep and `/vendor`
- klog

### Notes:
- Improved DNS ClusterIP calculation:
  - agnostic to IPv4/IPv6
  - works with subnets that do not start with .1
- Corrected `rbac.authorization.kubernetes.io/autoupdate` to be false
- `r.createOrFetch()` is a weird function (consequence of unrolling a generic loop on runtime.Object's to do type-specific updates)
- kubebuilder docs quickly mention to use patch instead of update for version reasons
  why and how?
- Should write some tests
- TODO: Add aggregated ClusterRoles for `admin` and `edit` matching the CoredDNS CRD ([example](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/deploy/upstream/manifests/0.9.0/0000_50_olm_09-aggregated.clusterrole.yaml))
  (does kubebuilder generate this?  operator-sdk does not)

### operator-sdk learnings:
- single binary / no dependencies  (kubebuilder needs modified `PATH`)
- opinionated build using docker or buildah is easy to use
- provided RBAC role was missing explicit definition for leader-election
- generated code similar to kubebuilder
- docs are good
- `operator-sdk test local` seems broken at the moment
- built-in support for operator Prometheus metrics /w ServiceMonitors -- did not try out
- some OLM(Operator Lifecycly Manager) commands integrated into CLI -- did not try out
- OLM appears to be useful for non-operator-sdk projects as well
- https://operatorhub.io/ may be a thoughtful place to publish to

### CODE REVIEW

**Comment by dholbach** [comment]
9299 added files is quite a bit. Does this all come from operator-sdk?

**Comment by stealthybox** [comment]
> 9299 added files is quite a bit. Does this all come from operator-sdk?

This is dep and /vendor
The Gopkg.toml is generated by operator-sdk

kubebuilder does the same thing, but @justinsb did the following to switch to go modules:
https://github.com/kubernetes-sigs/addon-operators/pull/7/files#diff-d2bbd32db0a3b86a0c4c75fbbfb68223R16
He chose not to check the /vendor in.

File: coredns-operator/pkg/controller/coredns/coredns_controller.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if instance.Spec.Corefile == "" {
+		instance.Spec.Corefile = DefaultCorefile
+		reqLogger.Info("Updating CoreDNS with default Corefile")
+		err = r.client.Update(context.TODO(), instance)
```
Comment (justinsb) [inline coredns-operator/pkg/controller/coredns/coredns_controller.go]: This is nice - this is effectively defaulting, IIUC

File: coredns-operator/pkg/controller/coredns/coredns_controller.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+				err = r.client.Delete(context.TODO(), foundService)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+			}
+		}
+	}
+
+	foundConfigMap := &corev1.ConfigMap{}
+	err = r.createOrFetch(reqLogger, instance, newConfigMapForCR(instance), foundConfigMap)
```
Comment (justinsb) [inline coredns-operator/pkg/controller/coredns/coredns_controller.go]: Do we have to do this?  I'm obviously biased, but I much prefer the declarative approach both from the point of view of not requiring recompilation, and from the point of view of creating something that we can realistically expect everyone to use.

Could we combine operator-sdk with the [declarative library](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern) ?  What additional things does operator-sdk add over kubebuilder if we do that?

**Comment by justinsb** [comment]
Thanks for doing this @stealthybox ... one thing I wasn't clear about is whether we _have_ to generate the objects from code, or whether there is functionality for sourcing from a yaml manifest (or similar).

Sourcing from a file makes for a smaller gap from `kubectl apply`, in my opinion, and it also means we don't need to create a new operator version for every addon version.

I'm actually surprised by how similar this has ended up to the kubebuilder code.  As I see it the primary difference is around declarative construction, and then use of apply or not.  But both approaches have custom CRDs, both generally leverage kubebuilder, etc.  So this makes me very optimistic!

(I do also want to move to go modules everywhere, but it's a bit of a project in and of itself.  But I think it's only a matter of time, anyway!)

File: coredns-operator/Gopkg.toml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+[[override]]
+  name = "github.com/coreos/prometheus-operator"
+  version = "=v0.29.0"
+
+[[override]]
+  name = "sigs.k8s.io/controller-runtime"
+  version = "=v0.1.10"
+
+[[constraint]]
+  name = "github.com/operator-framework/operator-sdk"
```
Comment (timothysc) [inline coredns-operator/Gopkg.toml]: I'm generally concerned about being bound to vendor controlled apis/concepts vs. leveraging project or community owned interfaces.   

For example we can influence kube-builder, but if there is an impass we may need to fork the operator framework.

Comment (timothysc) [reply coredns-operator/Gopkg.toml]: Also, this seems like an insane level of vendoring.

**Comment by dholbach** [comment]
I think the idea from last meeting was to take some of the extra functionality from this operator and merge it into https://github.com/kubernetes-sigs/addon-operators/pull/7 - correct?

**Comment by stealthybox** [comment]
> I think the idea from last meeting was to take some of the extra functionality from this operator and merge it into #7 - correct?

This is correct :+1:

**Pull request closed without merging**